src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,210 @@
+        }
+        Poll::Pending
+    }
+```
+
+Of course, there's something missing here. How does the callback wake the
+`PrintFuture`? She needs to store the
+[`Waker`](https://doc.rust-lang.org/stable/core/task/struct.Waker.html)
+somewhere! Barbara puts the `Waker` in a global variable so the callback can
+find it (this is fine because the app is single threaded and callbacks do NOT
+interrupt execution the way Unix signals do):
+
+```rust
+static mut PRINT_WAKER: Option<Waker> = None;
+
+extern fn callback() {
+    if let Some(waker) = unsafe { PRINT_WAKER.as_ref() } {
+        waker.wake_by_ref();
+    }
+}
+```
+
+She then modifies `poll` to set `PRINT_WAKER`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            return Poll::Ready(kernel_get_buffer_back());
+        }
+        unsafe { PRINT_WAKER = Some(cx.waker()); }
+        Poll::Pending
+    }
+```
+
+At this point, Barbara starts to think about the size impact of these changes.
+`PRINT_WAKER` is stored in `.bss`, which occupies space in RAM but not flash. It
+is two words in size. It points to a
+[`RawWakerVTable`](https://doc.rust-lang.org/stable/core/task/struct.RawWakerVTable.html)
+that is provided by the executor. Considering that this code is designed to run
+without `alloc`, `RawWakerVTable` seems awfully wasteful. `drop` is a no-op,
+`clone` seems like it will be a no-op, and `wake`/`wake_by_ref` seem like
+duplicates.
+
+At this point, Barbara thought about the overhead and decided it would be worth
+it if it made it easy to work with future combinators from external crates, and
+could function as a lightweight form of multithreading. To support this, Barbara
+decides to implement an executor designed to function like a background thread.
+This executor is started using a `spawn` function, and otherwise runs entirely
+in kernel callbacks.
+
+Barbara identifies several factors that add branching and error handling code to
+the executor:
+
+1. `spawn` should be a safe function, because it is called by high-level
+   application code. However, that means it can be called by the future it
+   contains. If handled naively, this would result in dropping the future while
+   it executes. Barbara adds runtime checks to identify this situation.
+1. `Waker` is `Sync`, so on a multithreaded system, a future could give another
+   thread access to its `Waker` and the other thread could wake it up. This
+   could happen while the `poll` is executing, before `poll` returns
+   `Poll::Pending`. Therefore, Barbara concludes that if `wake` is called while
+   a future is being polled then the future should be re-polled, even if the
+   current `poll` returns `Poll::Pending`. This requires putting a retry loop
+   into the executor.
+1. A kernel callback may call `Waker::wake` after its future finishes executing,
+   and the executor must ignore those wakeups. This duplicates the "ignore
```

Comment:
Can you add a sentence on *why* must the executor ignore the spurious wakeups?

Reply:
I reworded it to state that the executor should not `poll` a future after the future has returned `Poll::Ready`, which is directly stated in the `Future` docs. Is it clear now?